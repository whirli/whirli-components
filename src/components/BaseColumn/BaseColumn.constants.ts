export const PropKeys: Record<string, string> = {
  SPANS: 'spans',
  ALIGN: 'align',
  JUSTIFY: 'justify',
  DIRECTION: 'direction',
  START: 'start',
};

export const PropValues: Record<string, string[]> = {
  [PropKeys.SPANS]: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
  [PropKeys.ALIGN]: ['default', 'end', 'between'],
  [PropKeys.JUSTIFY]: ['default', 'end', 'between'],
  [PropKeys.DIRECTION]: ['default', 'row', 'row-reverse', 'column-reverse'],
  [PropKeys.START]: ['default', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12'],
};
