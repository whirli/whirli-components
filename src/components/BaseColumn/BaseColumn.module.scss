@import '@whirli-components/styles/Breakpoints.module.scss';

// default styles
.column {
  display: flex;
  flex-direction: column;
  overflow: auto;
  grid-column: span 12 / span 12;
}

.direction-default {
  flex-direction: column;
}

.direction-column-reverse {
  flex-direction: column-reverse;
}

.direction-row {
  flex-direction: row;
}

.direction-row-reverse {
  flex-direction: row-reverse;
}

// breakpoint styles
$packageBreakpointStyles: (
    spans-1: ( grid-column: span 1 / span 1 ),
    spans-2: ( grid-column: span 2 / span 2 ),
    spans-3: ( grid-column: span 3 / span 3 ),
    spans-4: ( grid-column: span 4 / span 4 ),
    spans-5: ( grid-column: span 5 / span 5 ),
    spans-6: ( grid-column: span 6 / span 6 ),
    spans-7: ( grid-column: span 7 / span 7 ),
    spans-8: ( grid-column: span 8 / span 8 ),
    spans-9: ( grid-column: span 9 / span 9 ),
    spans-10: ( grid-column: span 10 / span 10 ),
    spans-11: ( grid-column: span 11 / span 11 ),
    spans-12: ( grid-column: span 12 / span 12 ),

    align-default: (
      align-items: flex-start,
    ),
    align-end: (
      align-items: flex-end,
    ),
    align-center: (
      align-items: center,
    ),
    align-between: (
      align-items: space-between,
    ),

    justify-default: (
      justify-content: flex-start,
    ),
    justify-center: (
      justify-content: center,
    ),
    justify-end: (
      justify-content: flex-end,
    ),
    justify-between: (
      justify-content: space-between,
    ),

    start-default: ( grid-column-start: auto ),
    start-1: ( grid-column-start: 1 ),
    start-2: ( grid-column-start: 2 ),
    start-3: ( grid-column-start: 3 ),
    start-4: ( grid-column-start: 4 ),
    start-5: ( grid-column-start: 5 ),
    start-6: ( grid-column-start: 6 ),
    start-7: ( grid-column-start: 7 ),
    start-8: ( grid-column-start: 8 ),
    start-9: ( grid-column-start: 9 ),
    start-10: ( grid-column-start: 10 ),
    start-11: ( grid-column-start: 11 ),
    start-12: ( grid-column-start: 12 ),
);