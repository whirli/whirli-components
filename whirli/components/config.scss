@import '@whirli-components/styles/Helpers.scss';

$breakpointPrefix: BREAKPOINT_;

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 672px,
    ),
    small_desk: (
        prefix: sd,
        min-width: 1056px
    ),
    large_desk: (
        prefix: ld,
        min-width: 1312px
    ),
    huge_desk: (
        prefix: hd,
        min-width: 1584px
    ),
);

:export {
    prefix: $breakpointPrefix;
    #{$breakpointPrefix}default: 0px;
    @each $name, $value in $breakpoints {
        #{$breakpointPrefix}#{$name}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}