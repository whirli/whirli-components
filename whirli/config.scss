@import '@whirli-components/src/styles/Helpers.scss';

$breakpoints: (
    tablet: (
        prefix: t,
        min-width: 48rem,
    ),
    desktop: (
        prefix: d,
        min-width: 80rem,
    )
);

:export {
    @each $name, $value in $breakpoints {
        BREAKPOINT_#{to-upper-case(#{$name})}: map-deep-get($breakpoints, #{$name}, 'min-width')
    }
}