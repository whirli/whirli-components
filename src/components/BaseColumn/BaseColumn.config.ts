// @ts-ignore
import { PropType } from '@composition';
import * as PropHelpers from '@whirli-components/helpers/props';
import { ComponentStyles, ComponentProps } from '@whirli-components/@types/components';

// Local config
import config from '@whirli-local/components/BaseColumn/BaseColumn.config';

// Package config
import { PropKeys } from './BaseColumn.constants';
import type { PropSpans, PropAlign, PropJustify, PropDirection } from './BaseColumn.types';

export const ConfigStyles: ComponentStyles = {
  ...config.styles,
  [PropKeys.SPANS]: {
    generateBreakpoints: true,
    classes: {
      '1': 'span-1',
      '2': 'span-2',
      '3': 'span-3',
      '4': 'span-4',
      '5': 'span-5',
      '6': 'span-6',
      '7': 'span-7',
      '8': 'span-8',
      '9': 'span-9',
      '10': 'span-10',
      '11': 'span-11',
      '12': 'span-12',
      ...config.styles[PropKeys.SPANS]?.classes,
    } as Record<PropSpans, string>,
  },
  [PropKeys.ALIGN]: {
    generateBreakpoints: true,
    classes: {
      default: 'align-default',
      center: 'align-center',
      end: 'align-end',
      between: 'align-between',
      ...config.styles[PropKeys.ALIGN]?.classes,
    } as Record<PropAlign, string>,
  },
  [PropKeys.JUSTIFY]: {
    generateBreakpoints: true,
    classes: {
      default: 'justify-default',
      center: 'justify-center',
      end: 'justify-end',
      between: 'justify-between',
      ...config.styles[PropKeys.JUSTIFY]?.classes,
    } as Record<PropJustify, string>,
  },
  [PropKeys.DIRECTION]: {
    generateBreakpoints: true,
    classes: {
      default: 'direction-default',
      row: 'direction-row',
      'row-reverse': 'direction-row-reverse',
      ...config.styles[PropKeys.DIRECTION]?.classes,
    } as Record<PropDirection, string>,
  },
};

export const ConfigProps: ComponentProps = {
  [PropKeys.SPANS]: {
    required: false,
    type: [String, Object] as PropType<PropSpans | Record<string, PropSpans>>,
    default: '12',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.SPANS),
  },
  [PropKeys.ALIGN]: {
    required: false,
    type: [String, Object] as PropType<PropAlign | Record<string, PropAlign>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.ALIGN),
  },
  [PropKeys.JUSTIFY]: {
    required: false,
    type: [String, Object] as PropType<PropJustify | Record<string, PropJustify>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.JUSTIFY),
  },
  [PropKeys.DIRECTION]: {
    required: false,
    type: [String, Object] as PropType<PropDirection | Record<string, PropDirection>>,
    default: 'default',
    ...PropHelpers.breakpointStylePropValidation(ConfigStyles, PropKeys.DIRECTION),
  },
  ...config.props,
};
