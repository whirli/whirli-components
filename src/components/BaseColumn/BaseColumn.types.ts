import { PropKeys, PropValues } from './BaseColumn.constants';

const propSpans = PropValues[PropKeys.SPANS];
const propAlign = PropValues[PropKeys.ALIGN];
const propJustify = PropValues[PropKeys.JUSTIFY];
const propDirection = PropValues[PropKeys.DIRECTION];

export type PropSpans = typeof propSpans[number];
export type PropAlign = typeof propAlign[number];
export type PropJustify = typeof propJustify[number];
export type PropDirection = typeof propDirection[number];
