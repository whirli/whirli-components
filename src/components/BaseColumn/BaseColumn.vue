<template>
  <div :class="classes"><slot /></div>
</template>

<script setup lang="ts">
// Vue
// Styles
// @ts-ignore
import styles from '@whirli-local/components/BaseColumn/BaseColumn.module.scss?module';

// Data
import { ConfigStyles, ConfigProps } from './BaseColumn.config';

const ComponentStyles = ConfigStyles;

const props = defineProps(ConfigProps);

// Classes
import useClasses from '@whirli-components/@use/class';
const { makeClasses } = useClasses();
const classes = [styles.column, ...makeClasses(ComponentStyles, props, styles)];
</script>
