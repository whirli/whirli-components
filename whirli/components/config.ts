import BaseAside from './BaseAside/BaseAside.config';
import BaseButton from './BaseButton/BaseButton.config';
import BaseColumn from './BaseColumn/BaseColumn.config';
import BaseContainer from './BaseContainer/BaseContainer.config';
import BaseGrid from './BaseGrid/BaseGrid.config';
import BaseImage from './BaseImage/BaseImage.config';
import BaseMain from './BaseMain/BaseMain.config';
import BaseSection from './BaseSection/BaseSection.config';
import BaseSpacer from './BaseSpacer/BaseSpacer.config';
import BaseText from './BaseText/BaseText.config';

export default {
  BaseAside,
  BaseButton,
  BaseColumn,
  BaseContainer,
  BaseGrid,
  BaseImage,
  BaseMain,
  BaseSection,
  BaseSpacer,
  BaseText,
};
