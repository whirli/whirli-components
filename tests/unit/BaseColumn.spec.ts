import { shallowMount } from '@vue/test-utils';
import { propReturnsClass, propReturnsBreakpointClasses } from '../helpers';

import BaseColumn from '@/components/BaseColumn/BaseColumn.vue';
import { PropKeys } from '@/components/BaseColumn/BaseColumn.constants';
import { ConfigStyles } from '@/components/BaseColumn/BaseColumn.config';

let wrapper: any;

describe('BaseColumn.vue - Single classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseColumn, {
      props: {
        spans: '6',
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('maps to the correct styles', () => {
    expect(propReturnsClass(wrapper, ConfigStyles, PropKeys.SPANS, 'span-6')).toBe(true);
  });
});

describe('BaseColumn.vue - Breakpoint classes', () => {
  beforeEach(() => {
    wrapper = shallowMount(BaseColumn, {
      props: {
        spans: { default: '12', tablet: '6' },
      },
    });
  });

  it('renders the component', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('maps to the correct styles', () => {
    expect(
      propReturnsBreakpointClasses(wrapper, ConfigStyles, PropKeys.SPANS, {
        default: 'span-12',
        tablet: 'span-6',
      })
    ).toBe(true);
  });
});
