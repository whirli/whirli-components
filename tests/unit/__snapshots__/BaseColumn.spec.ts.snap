// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseColumn.vue - Breakpoint classes renders the component 1`] = `
<div
  class=""
>
  
  
</div>
`;

exports[`BaseColumn.vue - Single classes renders the component 1`] = `
<div
  class=""
>
  
  
</div>
`;
